evennia/contrib/slots_handler/slots.py

```diff
@@ -0,0 +1,385 @@
+"""
+Slots Handler
+
+This class is designed to sit on typeclassed objects and allow for other
+objects to be attached to configurable slots on the handler-endowed object.
+
+Built by DamnedScholar (https://github.com/damnedscholar)
+"""
+
+
+from collections import OrderedDict
+import evennia
+from evennia.utils.dbserialize import _SaverDict, _SaverList, _SaverSet
+
+
+class SlotsHandler:
+    """
+    Handler for the slots system. This handler is designed to be attached to
+    objects with a particular @lazy-property name, so it will be referred to as
+    `.slots`, though individual games can set that however they like.
+
+    Place on character or other typeclassed object:
+    ```
+    @lazy_property
+    def slots(self):
+        return SlotsHandler(self)
+    ```
+
+    The purpose of this handler is to sit on a "holder" typeclassed object and
+    manage slots for "held" objects. The first thing that should happen is
+    using `self.slots.add()` on the holder object to identify the available
+    options. `.slots.delete()` can be used to delete or pop old slots.
+
+    By default, a held object can have slots stored on it (the handler will
+    check `.db.slots`, `.ndb.slots`, and `.slots` in that order), and
+    `self.slots.attach()` will attempt to attach it on all of those slots (it
+    should fail if it can't attach on every slot). `self.slots.drop()` will
+    remove the target from any of the slots given it. Both of these methods
+    have the option to be given a custom `slots` argument, which will override
+    the slots on the held object.
+
+    Using the custom `slots` in `attach()` and `drop()` provides some
+    customization facility in that you can store *any* object in a slot, not
+    just an object that has been set up for it.
+    """
+
+    def __init__(self, obj):
+        self.obj = obj
+        self._objid = obj.id
+
+    # Internal utility methods
+    def is_name_valid(self, test):
+        "Internal function to check if the name is right."
+        try:
+            exec(test + " = 1")
+        except SyntaxError:
+            raise ValueError("You have to limit slot category names to "
+                             "characters that are valid in variable names.")
+
+    def defrag_nums(self, cats):
```

Comment:
Since this is not a private function, it should have a *Google style docstring* as per the [contribution guidelines](https://github.com/evennia/evennia/wiki/Contributing#contributing-with-contribs). This is true for all methods and functions in the contrib (I won't mark them each explicitly).